### ISSUE TITLE
Keyspace Events

### ISSUE THREAD

**Issue by whitejv81** [davidjrh/dnn.rediscachingprovider#5]
Here's the issue I'm seeing:  we have two servers if something is updated on one, the other does not get the update.  For example, if I move a module on the home page, I see a key written to Redis "TabModules55".  Viewing the server on which the content was not updated the module is still in the old position until the application pool is recycled or internal cache on this server is cleared by some other means.  Is this expected behavior or should the servers stay in sync using the Redis Keyspace events?  I'm thinking about adding this line to see if things start staying in sync, but feedback is welcome.  Maybe I have something configured incorrectly?

`cn.GetSubscriber().Subscribe(new RedisChannel(string.Format("__keyspace@0__:{0}*", KeyPrefix), RedisChannel.PatternMode.Pattern), ProcessMessage);`

**Comment by davidjrh** [comment]
Hi, definitively that is not the expected behavior. There are objectives for the Redis caching provider: 1)  act as a caching provider to minimize database roundtrips using an out-of-memory cache; 2) maintaing that cache in sync between the IIS nodes 

The second one is accomplished by implementing the Redis subscriber feature. The subscriber is created when the Connection is created: 

```c#
		private static readonly Lazy<ConnectionMultiplexer> LazyConnection = new Lazy<ConnectionMultiplexer>(() =>
		{
			var cn = ConnectionMultiplexer.Connect(ConnectionString);
			cn.GetSubscriber()
				.Subscribe(new RedisChannel(KeyPrefix + "Redis.*", RedisChannel.PatternMode.Pattern),
					ProcessMessage);
			return cn;
		});

		private static void ProcessMessage(RedisChannel redisChannel, RedisValue redisValue)
		{
			try
			{
				var instance = (RedisCachingProvider) Instance() ?? new RedisCachingProvider();
				if (redisChannel == KeyPrefix + "Redis.Clear")
				{				
					if (redisValue != InstanceUniqueId) // Avoid to clear twice
					{
						instance.Clear("", "", false);                        
					}
				}
				else
				{
				    if (redisValue.ToString().Length > InstanceUniqueId.Length &&
				        !redisValue.ToString().StartsWith(InstanceUniqueId))
				    {
						instance.Remove(redisValue.ToString().Substring(InstanceUniqueId.Length + 1), false);                     
				    }

				}
			}
			catch (Exception e)
			{
				if (!ProcessException(e)) throw;
			}
		}
```

Then the key updates/deletes are notified via publish

```c#
					// Notify the channel
					RedisCache.Publish(new RedisChannel(KeyPrefix + "Redis.Remove", RedisChannel.PatternMode.Auto), InstanceUniqueId + "_" + key);
```

I belive the problem can come from the "Insert" implementation, not publishing a "Redis.Remove" so the other nodes clear the in-memory cache object. I'm going to do some tests and see if that fixes the issue.

**Comment by whitejv81** [comment]
Good morning,

Thanks for looking into this.  I've done a bit more troubleshooting and I now finally understand the ProcessMessage function.  I added a lot more logging to step through the code to figure out what's happening.  I originally thought the Subscribers were not receiving the messages properly, but it turns out the code functions properly with the exception that the Subscribers are not actually clearing their cache.  Example:  moving a tab on the page issues a Publish(Redis.Clear).  The Subscribers enter the ProcessMessage and execute Clear(), however the changes made on Server1 do not appear on Server2. I added DataCache.ClearCache(); below the ClearCacheInternal() call.  This worked, but also intermittently caused an infinite loop on Redis (which was determined by subscribing to the Redis.* channel to see what messages were being passed around).  Admittedly, this is a big of a sledge hammer approach, but it seemed to keep things in sync.  I also tried to modify the Remove() method by using DataCache.RemoveCache(key).  This was a bad thing...  After the install was completed, it left the instance in a white screen of death until I modified the web.config to change the CachingProvider back to FileBasedCaching.

So, I guess to summarize, I think the ClearCacheInternal isn't doing what is expected, but I can't really prove it.  I'm quite new to developing for DNN and I feel a bit like I'm just throwing pasta against the wall to see what sticks.

**Comment by whitejv81** [comment]
A little more logging and this is where I'm at now:  moving a tab on a page causes a call to `ClearCacheInternal("Module", 55, true);`.  As far as I can tell, what should actually be getting nuked is the CacheKey TabModule55.

**Comment by whitejv81** [comment]
I subscribed to the keyspace for the cacheKey TabModules55 and here are the events that happen to that key when a tab is moved.
```
127.0.0.1:6379> subscribe '__keyspace@0__:redis_DNN_TabModules55'
Reading messages... (press Ctrl-C to quit)
1) "subscribe"
2) "__keyspace@0__:redis_DNN_TabModules55"
3) (integer) 1
1) "message"
2) "__keyspace@0__:redis_DNN_TabModules55"
3) "del"
1) "message"
2) "__keyspace@0__:redis_DNN_TabModules55"
3) "set"
1) "message"
2) "__keyspace@0__:redis_DNN_TabModules55"
3) "expire"
```

**Comment by davidjrh** [comment]
Hi, I found the issue causing the out of sync cache. On parametrized ClearCache calls, the "type" and "data" parameters were not being published on the Redis channel, so were not effectively being cleared on the other webservers. 

![image](https://cloud.githubusercontent.com/assets/489976/12212046/b80e11c8-b65d-11e5-9f34-a63e74212d06.png)


I have added those parameters and now is working properly.I have created a new release 1.0.4 with the fix.
![image](https://cloud.githubusercontent.com/assets/489976/12212071/e9a9685e-b65d-11e5-87cf-57e9004453c7.png)
